### PR TITLE
Handling Pausing from freezer state

### PR DESCRIPTION
**Pull request by rajasec**
@crosbymichael @mrunalp 
With current runc implementation, container pause is handled from the config file, Added the contents to handle from the freezer subsystem and its freezer.state from the respective container.
This is to handle #217 which has been raised

Signed-off-by: Rajasekaran <rajasec79@gmail.com>

### CODE REVIEW

**Comment by wking** [comment]
As it stands with 48dd4af, here's the current mapping from
freezer.state text [1] to Go constants:

* THAWED → Running
* FREEZING → Running
* FROZEN → Paused
* No freezer subsystem → Running

I'd suggest we update our Go constants to reflect the kernel's values:

* THAWED → Thawed
* FREEZING → Freezing
* FROZEN → Frozen
* No freezer subsystem → Thawed

so folks can distinguish between THAWED and FREEZING.  But I'm not
sure how other OSes handle this sort of state, so if we're trying to
strike a portable balance between their models, maybe Running / Paused
is the best we can do?  In that case, lumping FREEZING and THAWED
together is probably a lesser evil than lumping FREEZING and FROZEN
together, because trying to freeze a FREEZING cgroup is idempotent
[1], but trying to walk a FREEZING cgroup as if it were FROZEN could
give you an inconsistent picture.

[1]: https://www.kernel.org/doc/Documentation/cgroups/freezer-subsystem.txt

**Comment by rajasec** [comment]
@wking @crosbymichael @LK4D4
I agree with your point on consistency with respect to states and kernel values. Do you want me to update in this PR or in different PR ?

**Comment by wking** [comment]
On Fri, Aug 21, 2015 at 10:25:26AM -0700, Rajasekaran wrote:
> I agree with your point on consistency with respect to states and
> kernel values. Do you want me to update in this PR or in different
> PR ?

I'm not a maintainer (just adding my perspective as a user), so I
can't comment on how to split this among PRs.

Maybe this sort of model discussion should happen on the mailing list
[1]?  If 48dd4af is how Docker currently handles the mapping, then
it's probably fine to leave this PR as it stands until there's
something in the spec defining the intended semantics.

[1]: https://github.com/opencontainers/specs/tree/9ad789fa16156841074275172530fe7db6bcb26b#discuss-your-design

**Comment by crosbymichael** [comment]
@rajasec this looks good as a way to find out if the container is frozen.  

Lets factor out this code into a separate function and make sure that we have a specific error type if a system does not support the freezer cgroup.  We don't want to the state method to fail if a system does not support freezer.  So make sure you check for an IsNotExist error and ignore it in here.

**Comment by rajasec** [comment]
@crosbymichael 
Added the separate function to handle the freezer. Ensured, state method will not fail if system does not supporting freezer.
Latest commit 86c1d6f7e1e0efa8bbe5c9f3d52734937907668b

**Comment by rajasec** [comment]
@crosbymichael @wking 
Here is the latest commit as per your comments.
b73aef50f0594e3011fdea3010f3a3674efc42fb

**Comment by wking** [comment]
On Sat, Aug 22, 2015 at 12:14:09AM -0700, Rajasekaran wrote:
> Here is the latest commit as per your comments.
> b73aef50f0594e3011fdea3010f3a3674efc42fb

You can probably drop everything after your first s-o-b in your commit
message.

I've filed a fixup PR against this branch (rajasec/runc#1) with my
suggestion for dropping the racy Stat.

**Comment by wking** [comment]
On Sat, Aug 22, 2015 at 10:34:14AM -0700, W. Trevor King wrote:
> I've filed a fixup PR against this branch (rajasec/runc#1) with my
> suggestion for dropping the racy Stat.

I also added another fixup to that PR (wking/runc@6f3b1d2ea3) that
handles the case where the cgroups path mapping doesn't contain a
"freezer" key.  With your initial 48dd4afbef, you had a ‘path != ""’
check that might have been doing the same thing.  The GetPaths() docs
aren't clear [1], but I'd expect unavailable cgroups to just not be
listed in the mapping, rather than having a key with an empty-string
value.  That's how the filesystem-based manager works anyway [2].
It's unlikely that the empty string is a valid cgroup path, but it
seems better to use Go's builtin missing-key syntax than assuming that
the default string is always due to a missing-key lookup.

[1]: https://github.com/opencontainers/runc/blob/e7663a673efe0da696ea87eaa7695ec2e604dd76/libcontainer/cgroups/cgroups.go#L33-L35
[2]: https://github.com/opencontainers/runc/blob/e7663a673efe0da696ea87eaa7695ec2e604dd76/libcontainer/cgroups/fs/apply_raw.go#L112-L118

**Comment by rajasec** [comment]
@wking @crosbymichael 
I'm not checking for paths != "", With my last commit I changed to stat of the file ( whether subsystem exists). As empty path can also be valid cgroup path.
we can check cgroups.IsNotFound function would give whether subsystem exists.

**Comment by wking** [comment]
On Sat, Aug 22, 2015 at 08:17:04PM -0700, Rajasekaran wrote:
> I'm not checking for paths != "", With my last commit I changed to
> stat of the file ( whether subsystem exists).

Right, in [1] I was just referencing your original commit.

> As empty path can also be valid cgroup path.  we can check
> cgroups.IsNotFound function would give whether subsystem exists.

The explicit !ok check in wking/runc@6f3b1d2 (part of my fixup
rajasec/runc#1) is distinguishing between “we don't have a freezer
group” (in which case, there's no need to een attempt to open a file)
and “we have a freezer group at <path>” (in which case I try to open
<path>/freezer.state, without a Stat call because of
wking/runc@c07c601f, also in rajasec/runc#1).

With your current b73aef5, you don't distinguish between “we don't
have a freezer group” and “we have a freezer group with an
empty-string path”.  Conflating those two is probably not a big deal
(as I said in [1]), but I still think that my !ok check is the better
way to go (if nothing else, it saves a syscall in the no-freezer-group
case).

[1]: https://github.com/opencontainers/runc/pull/218#issuecomment-133733195

**Comment by rajasec** [comment]
@crosbymichael 
Your thoughts.. Do I need to make any further updates

File: libcontainer/container_linux.go [hunk 1]
```diff
@@ -767,6 +767,25 @@ func (c *linuxContainer) updateState(process parentProcess) error {
 	return json.NewEncoder(f).Encode(state)
 }
 
+func (c *linuxContainer) checkFreezer() (Status, error) {
+	path := c.cgroupManager.GetPaths()["freezer"]
+	if _, err := os.Stat(path); err != nil {
+		if os.IsNotExist(err) {
+			return Running, nil
+		}
```
Comment (crosbymichael) [inline libcontainer/container_linux.go]: I think you are missing a `return 0, err` here

File: libcontainer/container_linux.go [hunk 2]
```diff
@@ -767,6 +767,25 @@ func (c *linuxContainer) updateState(process parentProcess) error {
 	return json.NewEncoder(f).Encode(state)
 }
 
+func (c *linuxContainer) checkFreezer() (Status, error) {
+	path := c.cgroupManager.GetPaths()["freezer"]
+	if _, err := os.Stat(path); err != nil {
+		if os.IsNotExist(err) {
+			return Running, nil
```
Comment (crosbymichael) [inline libcontainer/container_linux.go]: this seems a little weird to have here.  Maybe a comment explaining why the status would be running if the error is an IsNotExist error.

**Comment by rajasec** [comment]
@crosbymichael 
2 things
1) If the freezer subsystem is not available part of container creation, we don't want to fail in this case. Container status will be running
2) func TestGetContainerState(t *testing.T)  part of unit testing uses the container only with memory subsystem, To handle this situation too, we return running. Via this test only, it helped me to figure out container being created without  cgroup freezer.

Mainly for the first point to handle system without freezer subsystem.

Even I thought that the same thing, it seems little odd for return Running.

**Comment by wking** [comment]
On Fri, Aug 28, 2015 at 08:49:17PM -0700, Rajasekaran wrote:
> 1) If the freezer subsystem is not available part of container
>    creation, we don't want to fail in this case. Container status
>    will be running
> 2) func TestGetContainerState(t *testing.T) part of unit testing
>    uses the container only with memory subsystem, To handle this
>    situation too, we return running. Via this test only, it helped
>    me to figure out container being created without cgroup freezer.

(2) is just an instance of (1), so the comment can be just:

  if os.IsNotExist(err) {
    // there is no freezer subsystem, so the application must be running
    return Running, nil

**Comment by rajasec** [comment]
@crosbymichael 
Added the comment and committed the file.

File: libcontainer/container_linux.go [hunk 3]
```diff
@@ -775,6 +775,27 @@ func (c *linuxContainer) updateState(process parentProcess) error {
...
+			// return the status as Running
+			return Running, nil
+		}
+	}
+	contents, err := ioutil.ReadFile(filepath.Join(path, "freezer.state"))
+	if err != nil {
+		return 0, newSystemError(err)
+	}
+	freezerstate := string(contents)
+	if strings.TrimSpace(freezerstate) == "FROZEN" {
```
Comment (hqhq) [inline libcontainer/container_linux.go]: Maybe we should check both `FROZEN` and `FREEZING` here, we have `pausing` defined but never used, it should be returned here.

File: libcontainer/container_linux.go [hunk 4]
```diff
@@ -775,6 +775,27 @@ func (c *linuxContainer) updateState(process parentProcess) error {
 	return json.NewEncoder(f).Encode(state)
 }
 
+func (c *linuxContainer) checkFreezer() (Status, error) {
+	path := c.cgroupManager.GetPaths()["freezer"]
+	if _, err := os.Stat(path); err != nil {
+		if os.IsNotExist(err) {
+			// Freezer subsystem is not available,Application is neither Checkpointed nor Destroyed
+			// return the status as Running
```
Comment (hqhq) [inline libcontainer/container_linux.go]: The commend seems a bit odd to me, isn't `Checkpointed` already checked before this function is called?

**Comment by rajasec** [comment]
@hqhq 
Checkpointed is already checked before entering this function  ( currentStatus function had the immediate check on checkpoint. At the last checkFreezer is called.

File: libcontainer/container_linux.go (hunk 3 above)
Comment (wking) [inline libcontainer/container_linux.go]: On Mon, Sep 07, 2015 at 12:01:35AM -0700, Qiang Huang wrote:
> Maybe we should check both `FROZEN` and `FREEZING` here, we have
> `pausing` defined but never used, it should be returned here.

Previous thoughts on model state settings in [1,2,3].  But if we
already have state constants for each of Linux's three settings, I
think it makes sense to use them (at least until the runtime-agnostic
spec has something to say about freezing).

[1]: https://github.com/opencontainers/runc/pull/218#issuecomment-133487383
[2]: https://github.com/opencontainers/runc/pull/218#issuecomment-133503694
[3]: https://github.com/opencontainers/runc/pull/218#issuecomment-133520061

**Comment by rajasec** [comment]
@crosbymichael 
Do I need to make any modifications ?

**Comment by crosbymichael** [comment]
@rajasec i cherry picked your commit into #311 so now we have robust freezer checking for the states.  Thanks!

**Pull request closed without merging**